lib/core/include/scipp/core/dict.h

```diff
@@ -0,0 +1,405 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Jan-Lukas Wynen
+#pragma once
+
+#include <functional>
+#include <vector>
+
+#include "scipp/common/index.h"
+
+#include "scipp/core/except.h"
+
+namespace scipp::core::dict_detail {
+template <class It1, class It2> struct ValueType {
+  using type = std::pair<typename It1::value_type, typename It2::value_type>;
+};
+
+template <class It1> struct ValueType<It1, void> {
+  using type = typename It1::value_type;
+};
+
+template <class It1, class It2> struct ReferenceType {
+  using type = std::add_rvalue_reference_t<
+      std::pair<typename It1::reference, typename It2::reference>>;
+};
+
+template <class It1> struct ReferenceType<It1, void> {
+  using type = typename It1::reference;
+};
+
+template <class Container, class It1, class It2, size_t... IteratorIndices>
+class Iterator;
+
+template <class BaseIterator, class Func> class TransformIterator;
+
+template <class Container, class It1, class It2, size_t... IteratorIndices>
+class Iterator {
+  static_assert(sizeof...(IteratorIndices) > 0 &&
+                sizeof...(IteratorIndices) < 3);
+
+public:
+  using difference_type = std::ptrdiff_t;
+  using value_type = typename ValueType<It1, It2>::type;
+  using pointer = std::add_pointer_t<std::remove_reference_t<value_type>>;
+  using reference = typename ReferenceType<It1, It2>::type;
+
+  template <class T>
+  Iterator(std::reference_wrapper<Container> container, T &&it1)
+      : m_iterators{std::forward<T>(it1)}, m_container(container),
+        m_end_address(container.get().data() + container.get().size()) {}
+
+  template <class T, class U>
+  Iterator(std::reference_wrapper<Container> container, T &&it1, U &&it2)
+      : m_iterators{std::forward<T>(it1), std::forward<U>(it2)},
+        m_container(container),
+        m_end_address(container.get().data() + container.get().size()) {}
+
+  Iterator(const Iterator &other) = default;
+  Iterator(Iterator &&other) noexcept = default;
+  Iterator &operator=(const Iterator &other) = default;
+  Iterator &operator=(Iterator &&other) noexcept = default;
+  ~Iterator() noexcept = default;
+
+  decltype(auto) operator*() const {
+    expect_container_unchanged();
+    if constexpr (sizeof...(IteratorIndices) == 1) {
+      return *std::get<0>(m_iterators);
+    } else {
+      return std::make_pair(std::cref(*std::get<0>(m_iterators)),
+                            std::ref(*std::get<1>(m_iterators)));
```

Comment:
Remind me: I think we had decided that ideally we want to return `std::pair<const Key, Value> &`. But we cannot *store* `std::pair<const Key, Value>`. Can we return the former if we have the latter? If I remember correctly it is not? Can we force it using a cast?

Reply:
> Can we return the former if we have the latter?

Yes. But we cannot store it because `std::vector` needs to move its elements (at the very least in `erase`) which requires non-const types.

> Can we force it using a cast?

I guess with a `reinterpret_cast`. But I don't know if this can have bad side-effects.